File: Client-Python/command_processor.py
import sys
import time
import grpc

from typing import List

class CommandProcessor:
    SPACE = " "
    BGN_TUPLE = "<"
    END_TUPLE = ">"
    PUT = "put"
    READ = "read"
    TAKE = "take"
    EXIT = "exit"
    SLEEP = "sleep"
    GET_TUPLE_SPACES_STATE = "getTupleSpacesState"
    INVALID_SERVER_INPUT_ERROR = "Invalid response from server."
    OPERATION_FAILED = "operation failed:"

    def __init__(self, client_service):
        self.client_service = client_service

    def parse_input(self):
        exit_flag = False
        while not exit_flag:
            try:
                line = input("> ").strip()
                split = line.split(self.SPACE)
                command = split[0]

                if command == self.PUT:
                    self.put(split)
                elif command == self.READ:
                    self.read(split)
                elif command == self.TAKE:
                    self.take(split)
                elif command == self.GET_TUPLE_SPACES_STATE:
                    self.get_tuple_spaces_state()
                elif command == self.SLEEP:
                    self.sleep(split)
                elif command == self.EXIT:
                    exit_flag = True
                else:
                    self.print_usage()
                
            except EOFError:
                break

    def put(self, split: List[str]):
        # check if the input is valid
        if not self.input_is_valid(split):
            self.print_usage()
            return

        # get the tuple
        tuple_data = split[1]
        try:
            # put the tuple
            self.client_service.put(tuple_data)
            print()
        except grpc.RpcError as e:  # Thrown when the server encounters an error and responds with an exception
            print(f"{self.PUT} {self.OPERATION_FAILED} {e.details()}\n", file=sys.stderr)

    def read(self, split: List[str]):
        # check if the input is valid
        if not self.input_is_valid(split):
            self.print_usage()
            return

        # get the tuple
        tuple_data = split[1]
        try:
            # read the tuple
            result = self.client_service.read(tuple_data)

            if not self.tuple_is_valid(result):
                print(f"{self.READ} {self.OPERATION_FAILED} {self.INVALID_SERVER_INPUT_ERROR}", file=sys.stderr)
                return
            
            print(f"{result}\n")                
        except grpc.RpcError as e:  # Thrown when the server encounters an error and responds with an exception
            print(f"{self.READ} {self.OPERATION_FAILED} {e.details()}\n", file=sys.stderr)

    def take(self, split: List[str]):
        # check if the input is valid
        if not self.input_is_valid(split):
            self.print_usage()
            return

        # get the tuple
        tuple_data = split[1]
        try:
            # take the tuple
            result = self.client_service.take(tuple_data)

            if not self.tuple_is_valid(result):
                print(f"{self.TAKE} {self.OPERATION_FAILED} {self.INVALID_SERVER_INPUT_ERROR}", file=sys.stderr)
                return
            
            print(f"{result}\n")              
        except grpc.RpcError as e:  # Thrown when the server encounters an error and responds with an exception
            print(f"{self.TAKE} {self.OPERATION_FAILED} {e.details()}\n", file=sys.stderr)

    def get_tuple_spaces_state(self):
        result = []
        
        try:
            # get the tuple spaces state
            result = self.client_service.get_tuple_spaces_state()

            for tuple in result:
                if not self.tuple_is_valid(tuple):
                    print(f"{self.GET_TUPLE_SPACES_STATE} {self.OPERATION_FAILED} {self.INVALID_SERVER_INPUT_ERROR}", file=sys.stderr)
                    return
                
            # Print tuples in the desired format
            formatted_tuples = ', '.join(result)
            print(f"[{formatted_tuples}]\n")    
        except grpc.RpcError as e:   # Handle possible IO errors
            print(f"{self.TAKE} {self.OPERATION_FAILED} {e.details()}\n", file=sys.stderr)
        
    def sleep(self, split: List[str]):
        
        
        if len(split) != 2:
            self.print_usage()
            return

        try:
            sleep_time = int(split[1])
            self.client_service.debug("Client " + self.client_service.client_id + \
                                      "-> sleeping for " + sleep_time + " seconds.")
        except ValueError:
            self.print_usage()
            return
        try:
            time.sleep(sleep_time)
        except KeyboardInterrupt:
            raise RuntimeError("Sleep interrupted")

    def print_usage(self):
        print("Usage:\n"
              "- put <element[,more_elements]>\n"
              "- read <element[,more_elements]>\n"
              "- take <element[,more_elements]>\n"
              "- getTupleSpacesState\n"
              "- sleep <integer>\n"
              "- exit\n")

    def input_is_valid(self, input_data: List[str]) -> bool:
        if (len(input_data) < 2
                or not input_data[1].startswith(self.BGN_TUPLE)
                or not input_data[1].endswith(self.END_TUPLE)
                or len(input_data) > 2):
            return False
        return True
    
    def tuple_is_valid(self, input: str) -> bool:
        split = input.split(self.SPACE)

        if (len(split) != 1
                or not split[0].startswith(self.BGN_TUPLE)
                or not split[0].endswith(self.END_TUPLE)):
            return False
        return True

File: Client-Java/src/main/java/pt/ulisboa/tecnico/tuplespaces/client/CommandProcessor.java
package pt.ulisboa.tecnico.tuplespaces.client;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

import io.grpc.StatusRuntimeException;
import pt.ulisboa.tecnico.tuplespaces.client.grpc.ClientService;


public class CommandProcessor {

    private static final String SPACE = " ";
    private static final String BGN_TUPLE = "<";
    private static final String END_TUPLE = ">";
    private static final String PUT = "put";
    private static final String READ = "read";
    private static final String TAKE = "take";
    private static final String SLEEP = "sleep";
    private static final String EXIT = "exit";
    private static final String GET_TUPLE_SPACES_STATE = "getTupleSpacesState";
    private static final String INVALID_SERVER_INPUT_ERROR = "Invalid response from server.";
    private static final String OPERATION_FAILED = " operation failed: ";

    private final ClientService clientService;

    public CommandProcessor(ClientService clientService) {
        this.clientService = clientService;
    }

    void parseInput() {

        Scanner scanner = new Scanner(System.in);
        boolean exit = false;

        while (!exit) {
            System.out.print("> ");
            String line = scanner.nextLine().trim();
            String[] split = line.split(SPACE);
             switch (split[0]) {
                case PUT:
                    this.put(split);
                    break;

                case READ:
                    this.read(split);
                    break;

                case TAKE:
                    this.take(split);
                    break;

                case GET_TUPLE_SPACES_STATE:
                    this.getTupleSpacesState();
                    break;

                case SLEEP:
                    this.sleep(split);
                    break;

                case EXIT:
                    exit = true;
                    break;

                default:
                    this.printUsage();
                    break;
             }
        }
        scanner.close();
    }

    private String[] parseDelays(String[] input) {
        // Default delay values (assume "" if not provided)
        String[] delays = new String[] { "", "", "" };

        // Fill available delays
        for (int i = 2; i < input.length && i < 5; i++) {
            delays[i - 2] = input[i];
        }

        return delays;
    }

    private void put(String[] split) {

        // check if input is valid
        if (!this.inputIsValid(split)) {
            this.printUsage();
            return;
        }
        
        // get the tuple
        String tuple = split[1];

        String[] delays = parseDelays(split);

        // get the arguments after the tuple
        String delayReplica1 = delays[0];
        String delayReplica2 = delays[1];
        String delayReplica3 = delays[2];

        try {
            // put the tuple
            this.clientService.put(tuple, delayReplica1, delayReplica2, delayReplica3);
            System.out.println();
        } catch (StatusRuntimeException e) { // Thrown when the server encounters an error and responds with an exception
            System.err.println(PUT + OPERATION_FAILED + e.getStatus().getDescription() + "\n");
        }
    }

    private void read(String[] split){
        // check if input is valid
        if (!this.inputIsValid(split)) {
            this.printUsage();
            return;
        }
        
        // get the tuple
        String tuple = split[1];

        String[] delays = parseDelays(split);

        // get the arguments after the tuple
        String delayReplica1 = delays[0];
        String delayReplica2 = delays[1];
        String delayReplica3 = delays[2];

        String result = "";

        try {
            // read the tuple
            result = this.clientService.read(tuple, delayReplica1, delayReplica2, delayReplica3);
            // Making sure the server didn't return an invalid tuple
            if (!this.tupleIsValid(result)) {
                System.out.println(READ + OPERATION_FAILED + INVALID_SERVER_INPUT_ERROR + "\n");
                return;
            }
            System.out.println(result + "\n");
        } catch (StatusRuntimeException e) { // Thrown when the server encounters an error and responds with an exception
            System.err.println(READ + OPERATION_FAILED + e.getStatus().getDescription() + "\n");
        }
    }


    private void take(String[] split){
         // check if input is valid
        if (!this.inputIsValid(split)) {
            this.printUsage();
            return;
        }
        
        // get the tuple
        String tuple = split[1];

        String[] delays = parseDelays(split);

        // get the arguments after the tuple
        String delayReplica1 = delays[0];
        String delayReplica2 = delays[1];
        String delayReplica3 = delays[2];
        
        String result = "";

        try {
            // take the tuple
            result = this.clientService.take(tuple, delayReplica1, delayReplica2, delayReplica3);
            // Making sure the server didn't return an invalid tuple
            if (!this.tupleIsValid(result)) {
                System.out.println(TAKE + OPERATION_FAILED + INVALID_SERVER_INPUT_ERROR + "\n");
                return;
            }
            System.out.println(result + "\n");
        } catch (StatusRuntimeException e) { // Thrown when the server encounters an error and responds with an exception
            System.err.println(TAKE + OPERATION_FAILED + e.getStatus().getDescription() + "\n");
        }
    }

    private void getTupleSpacesState(){

        List<String> result = new ArrayList<>();

        try {
            // get the tuple spaces state
            result = this.clientService.getTupleSpacesState();
            // Making sure the server didn't return invalid tuples
            for (String tuple : result) {
                if (!this.tupleIsValid(tuple)) {
                    System.out.println(GET_TUPLE_SPACES_STATE + OPERATION_FAILED + INVALID_SERVER_INPUT_ERROR + "\n");
                    return;
                }
            }
            System.out.println(result + "\n"); 
        } catch (StatusRuntimeException e) { // Handle possible IO errors
            System.err.println(GET_TUPLE_SPACES_STATE + OPERATION_FAILED + e.getStatus().getDescription() + "\n");
        }
    }

    private void sleep(String[] split) {
      if (split.length != 2){
        this.printUsage();
        return;
      }
      Integer time;

      // checks if input String can be parsed as an Integer
      try {
         time = Integer.parseInt(split[1]);
      } catch (NumberFormatException e) {
        this.printUsage();
        return;
      }

      try {
        this.clientService.debug("Client " + this.clientService.getClientId() + 
                                      "-> sleeping for " + time + " seconds.");
        Thread.sleep(time*1000);
      } catch (InterruptedException e) {
        throw new RuntimeException(e);
      }
    }

    private void printUsage() {
        System.out.println("Usage:\n" +
                "- put <element[,more_elements]>\n" +
                "- read <element[,more_elements]>\n" +
                "- take <element[,more_elements]>\n" +
                "- getTupleSpacesState <server>\n" +
                "- sleep <integer>\n" +
                "- exit\n");
    }

    private boolean inputIsValid(String[] input){
        if (input.length < 2 || !input[1].substring(0,1).equals(BGN_TUPLE)
            || !input[1].endsWith(END_TUPLE)) {
            return false;
        }

        // Number of optional delay values
        int numDelays = input.length - 2;

        // Only accept 0 or exactly 3 delay values
        if (numDelays != 0 && numDelays != 3) {
            return false;
        }

        // Check if all delay values (if present) are non-negative integers
        for (int i = 2; i < input.length; i++) {
            if (!input[i].matches("\\d+")) {
                return false;
            }
        }
        return true;
    }

    private boolean tupleIsValid(String input){
        String[] split = input.split(SPACE);

        if (split.length != 1
            ||
            !split[0].substring(0,1).equals(BGN_TUPLE) 
            || 
            !split[0].endsWith(END_TUPLE)) {
            return false;
        }
        else {
            return true;
        }
    }
}
